From 1acfa61ccc1211223ed1cddaa16e6237141665c8 Mon Sep 17 00:00:00 2001
From: Pantelis Antoniou <pantelis.antoniou@konsulko.com>
Date: Thu, 26 Nov 2015 19:02:20 +0200
Subject: [PATCH] resindvd: seek: Fallback to time seek if sector fails

Some badly authored DVDs do not have a timemap so it's impossible
to seek using the sector method. Switch to the (less accurate)
libdvdnav time method if that's the case.

Signed-off-by: Pantelis Antoniou <pantelis.antoniou@konsulko.com>
---
 ext/resindvd/resindvdsrc.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

Index: gst-plugins-bad-1.4.5/ext/resindvd/resindvdsrc.c
===================================================================
--- gst-plugins-bad-1.4.5.orig/ext/resindvd/resindvdsrc.c
+++ gst-plugins-bad-1.4.5/ext/resindvd/resindvdsrc.c
@@ -2809,11 +2809,13 @@ rsn_dvdsrc_seek_to_time (resinDvdSrc * s
       GST_TIME_ARGS (ts));
 
   sector = rsn_dvdsrc_get_sector_from_time_tmap (src, ts);
-  if (sector < 0)
-    return FALSE;
-
+  if (sector < 0) {
+    GST_DEBUG_OBJECT (src, "Cannot get sector to seek to; switch to dvdnav_time_search\n");
+    res = dvdnav_time_search(src->dvdnav, GSTTIME_TO_MPEGTIME(ts));
+  } else {
+    res = dvdnav_sector_search (src->dvdnav, sector, SEEK_SET);
+  }
   src->discont = TRUE;
-  res = dvdnav_sector_search (src->dvdnav, sector, SEEK_SET);
 
   if (res != DVDNAV_STATUS_OK)
     return FALSE;
